kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,717 @@
+            cluster_yml = os.path.join(config.KS_KUBESPRAY_PATH, "cluster.yml")
+            if not os.access(cluster_yml, os.R_OK):
+                raise ValueError("KS_KUBESPRAY_PATH is not properly configured")
+            OpenStack.connection_status(kwargs)
+            return config.PROVISIONER_OK_STATE
+        except Exception as e:
+            logging.exception("Error engine status: %s", e)
+            return config.PROVISIONER_ERROR_STATE
+
+
+class Kubespray:
+    """Kubespray wrapper.
+
+    This approach is not scalable. It may be solved by storing ssh
+    keys in db and running ansible on the one of the master nodes.
+
+    :param str cluster_id:
+    :param str ssh_username:
+    :param str clusters_path:
+    :param str kubespray_path:
+    :param dict os_kwargs:
+
+    """
+
+    def __init__(self, *, cluster_id, ssh_username,
+                 clusters_path, kubespray_path, os_kwargs):
+        self.cluster_id = cluster_id
+        self.ssh_username = ssh_username
+        self.clusters_path = clusters_path
+        self.kubespray_path = kubespray_path
+        self.os_kwargs = os_kwargs
+        self.ssh_common_args = ("-o", "UserKnownHostsFile=/dev/null",
+                                "-o", "StrictHostKeyChecking=no",
+                                "-i", os.path.join(clusters_path, "ssh_key"))
+        self._make_files_dir()
+
+    def deploy(self, resources):
+        inventory = self._generate_inventory(resources)
+        self._save_inventory(inventory, "hosts.json")
+        self._create_group_vars()
+        self._wait_for_ping()
+        self._run_ansible()
+        return self._get_kubeconfig(resources["masters"][0]["ip"])
+
+    def scale(self, resources):
+        inventory = self._generate_inventory(resources)
+        self._save_inventory(inventory, "hosts.json")
+        self._wait_for_ping()
+        self._run_ansible(playbook="scale.yml")
+
+    def shrink(self, resources, *, new_slave_count):
+        hostnames = [s["hostname"] for s in resources["slaves"]]
+        hostnames.sort()
+        slaves_left = hostnames[:new_slave_count]
+        inv = self._generate_inventory(resources, keep_slaves=slaves_left)
+        self._save_inventory(inv, "remove.json")
+        self._run_ansible(playbook="remove-node.yml", inventory="remove.json")
+        return hostnames[new_slave_count:]
+
+    def delete(self):
+        shutil.rmtree(self._get_cluster_path())
+
+    def _save_inventory(self, inventory, filename):
+        with open(self._get_cluster_path(filename), "w") as fp:
+            json.dump(inventory, fp, indent=4)
+
+    def _create_group_vars(self):
+        src = os.path.join(self.kubespray_path, "inventory/sample/group_vars")
+        dst = self._get_cluster_path("group_vars")
+        shutil.copytree(src, dst)
+        with open(os.path.join(dst, "all.yml"), "a") as all_yaml:
+            all_yaml.write("\ncloud_provider: openstack\n")
```

Comment:
should we use yaml module here?

Reply:
Yes. I'll do this in next PR when fixing volumes.